src/main/java/com/glencoesoftware/omero/ms/image/region/ImageRegionVerticle.java

```diff
@@ -138,45 +292,311 @@ private void renderImageRegion(Message<String> message) {
+                byte[] imageRegion = request.get().execute(
+                        requestHandler::renderImageRegion);
+                if (imageRegion == null) {
+                    throw new ObjectNotFoundException("Cannot render region");
+                }
+                // Cache the image region
+                JsonObject setMessage = new JsonObject();
+                setMessage.put("key", imageRegionCtx.cacheKey());
+                setMessage.put("value", imageRegion);
+                vertx.eventBus().send(
+                        RedisCacheVerticle.REDIS_CACHE_SET_EVENT,
+                        setMessage);
+                future.complete(imageRegion);
+            } catch (Exception e) {
+                future.fail(e);
+            }
+        }, future);
+
+        timer.observeDuration();
+        return future;
+    }
+
+    /**
+     * Get cached image region if available or retrieve it from the server.
+     * @param imageRegionCtx request context
+     * @param request OMERO request based on the current context
+     * @return Future which will be completed with the image region byte array.
+     */
+    private Future<byte[]> getCachedImageRegion(
+            ImageRegionCtx imageRegionCtx, Supplier<OmeroRequest> request) {
+        Summary.Timer timer = getCachedImageRegionSummary.startTimer();
+        Future<byte[]> future = Future.future();
+
+        String key = imageRegionCtx.cacheKey();
+        vertx.eventBus().<byte[]>send(
+                RedisCacheVerticle.REDIS_CACHE_GET_EVENT, key, result -> {
+            try {
+                byte[] imageRegion =
+                        result.succeeded()? result.result().body() : null;
+                ImageRegionRequestHandler requestHandler =
+                        new ImageRegionRequestHandler(
+                                imageRegionCtx, context, families,
+                                renderingModels, lutProvider);
+                if (imageRegion != null) {
+                    Future<Boolean> step1 =
+                            canRead(imageRegionCtx, request, requestHandler);
+
+                    step1.compose(canRead -> {
+                        if (canRead) {
+                            log.info("Cache HIT {}", key);
+                            imageRegionCacheHit.inc();
+                            future.complete(imageRegion);
+                        } else {
+                            log.info("Cache MISS {}", key);
+                            imageRegionCacheMiss.inc();
+                            future.complete(null);
+                        }
+                    }, future);
+                } else {
+                    log.info("Cache MISS {}", key);
+                    imageRegionCacheMiss.inc();
+                    future.complete(null);
+                }
+            } catch (Exception e) {
+                future.fail(e);
+            }
+        });
+
+        timer.observeDuration();
```

Comment:
With all the async code above, observing the duration here won't tell us anything, it will end up being 0ms.